libeos-payg/manager.c

```diff
@@ -938,15 +974,35 @@ epg_manager_init_async (GAsyncInitable      *initable,
 
   EpgManager *self = EPG_MANAGER (initable);
   g_autoptr(GTask) task = g_task_new (self, cancellable, callback, user_data);
+  g_autoptr(GFile) expiry_time_file = get_expiry_time_file (self);
+
   g_task_set_source_tag (task, epg_manager_init_async);
   g_task_set_priority (task, priority);
   epg_multi_task_attach (task, 4);
 
-  /* Load the expiry time. */
-  g_autoptr(GFile) expiry_time_file = get_expiry_time_file (self);
+  /* Technically there's a race here since the file could be created right after
+   * we see that it doesn't exist, but it shouldn't matter since eos-paygd
+   * should be the only one modifying its state directory */
+  if (g_file_query_exists (expiry_time_file, cancellable))
+    {
+      /* Load the expiry time (deprecated but kept for backwards compat). */
+      g_file_load_contents_async (expiry_time_file, cancellable,
+                                  file_load_cb, g_object_ref (task));
+    }
+  else
+    {
+      /* Load the wall clock time of the last state save. */
+      g_autoptr(GFile) wallclock_time_file = get_wallclock_time_file (self);
 
-  g_file_load_contents_async (expiry_time_file, cancellable,
-                              file_load_cb, g_object_ref (task));
+      g_file_load_contents_async (wallclock_time_file, cancellable,
+                                  file_load_cb, g_object_ref (task));
+
+      /* Load the expiry seconds. */
+      g_autoptr(GFile) expiry_seconds_file = get_expiry_seconds_file (self);
+
+      g_file_load_contents_async (expiry_seconds_file, cancellable,
+                                  file_load_cb, g_object_ref (task));
+    }
```

Comment:
I’d do this the other way round, and try loading the wallclock time file first. If that fails, fall back to the expiry time file. Don’t use `g_file_query_exists()` at all — just handle `G_IO_ERROR_NOT_FOUND` from the file loading functions. That saves on syscalls as well as eliminating the race condition.

Reply:
You're right, that seems better. It's slightly unfortunate that it means we can't have a clean separation between loading files in the `init_async` and processing them in the callback, but that's a small price to pay.